pkg/kubelet/images/image_pull_manager.go

```diff
@@ -0,0 +1,663 @@
+		return err
+	}
+
+	f.intentCounters[image]++
+	return nil
+}
+
+func (f *FileBasedImagePullManager) RecordImagePulled(image, imageRef string, credentials *imagemanagerv1alpha1.ImagePullCredentials) {
+	if err := f.writePulledRecord(image, imageRef, credentials); err != nil {
+		klog.ErrorS(err, "failed to write image pulled record", "imageRef", imageRef)
+		return
+	}
+
+	f.dereferenceImagePullIntent(image)
+}
+
+// writePulledRecord writes an ImagePulledRecord into the f.pulledDir directory.
+func (f *FileBasedImagePullManager) writePulledRecord(image, imageRef string, credentials *imagemanagerv1alpha1.ImagePullCredentials) error {
+	f.pulledAccessors.Lock(imageRef)
+	defer f.pulledAccessors.Unlock(imageRef)
+
+	var pulledRecord imagemanagerv1alpha1.ImagePulledRecord
+	var pulledRecordChanged bool
+
+	sanitizedImage, err := trimImageTagDigest(image)
+	if err != nil {
+		return fmt.Errorf("invalidate image name %q: %w", image, err)
+	}
+
+	pulledRecordPath := filepath.Join(f.pulledDir, cacheFilename(imageRef))
+	pulledFile, err := os.ReadFile(pulledRecordPath)
+	if os.IsNotExist(err) {
+		pulledRecordChanged = true
+		pulledRecord = imagemanagerv1alpha1.ImagePulledRecord{
+			LastUpdatedTime:   metav1.Time{Time: time.Now()},
+			ImageRef:          imageRef,
+			CredentialMapping: make(map[string]imagemanagerv1alpha1.ImagePullCredentials),
+		}
+		if credentials != nil {
+			pulledRecord.CredentialMapping[sanitizedImage] = *credentials
+		}
+	} else if err == nil {
+		if _, _, err := f.decoder.Decode(pulledFile, nil, &pulledRecord); err != nil {
+			return fmt.Errorf("failed to decode ImagePulledRecord: %w", err)
+		}
+
+		pulledRecord, pulledRecordChanged = pulledRecordMergeNewCreds(&pulledRecord, sanitizedImage, credentials)
+	} else {
+		return fmt.Errorf("failed to open %q for reading: %w", pulledRecordPath, err)
+	}
+
+	if !pulledRecordChanged {
+		return nil
+	}
+	recordBytes := bytes.NewBuffer([]byte{})
+	if err := f.encoder.Encode(&pulledRecord, recordBytes); err != nil {
+		return fmt.Errorf("failed to serialize ImagePulledRecord: %w", err)
+	}
+
+	if err := writeFile(f.pulledDir, cacheFilename(imageRef), recordBytes.Bytes()); err != nil {
+		return fmt.Errorf("failed to write ImagePulledRecord file: %w", err)
+	}
+
+	return nil
+
+}
+
+func (f *FileBasedImagePullManager) RecordImagePullFailed(image string) {
+	f.dereferenceImagePullIntent(image)
+}
+
+// dereferenceImagePullIntent decreses the number of how many times image pull
+// intent for a given `image` was requested, and removes the ImagePullIntent file
+// if the reference counter for the image reaches zero.
+func (f *FileBasedImagePullManager) dereferenceImagePullIntent(image string) {
+	f.intentAccessors.Lock(image)
+	defer f.intentAccessors.Unlock(image)
+
+	f.intentCounters[image]--
+	if f.intentCounters[image] > 0 {
+		return
+	}
+	f.intentCounters[image] = 0 // safeguard, we should never actually go below zero
```

Comment:
drop instead of accumulate zero values?

```suggestion
	delete(f.intentCounters, image)
```